File: design_pattern/solid/isp/solid_isp.go
package main

/// ISP (Interface Segregation Principle)
/// 1. 大きなインターフェースを小さなインターフェースに分割する
/// 2. クライアントが利用しないメソッドに依存しないようにする

type Document struct {
}

// ISP違反
type Machine interface {
	Print(d Document)
	Fax(d Document)
	Scan(d Document)
}

// 複数のデバイスが必要な場合
type MultiFunctionPrinter struct {
	// ...
}

func (m MultiFunctionPrinter) Print(d Document) {

}

func (m MultiFunctionPrinter) Fax(d Document) {

}

func (m MultiFunctionPrinter) Scan(d Document) {

}

type OldFashionedPrinter struct {
	// ...
}

func (o OldFashionedPrinter) Print(d Document) {
	// ok
}

func (o OldFashionedPrinter) Fax(d Document) {
	panic("operation not supported")
}

// Deprecated: ...
func (o OldFashionedPrinter) Scan(d Document) {
	panic("operation not supported")
}

// LSPに従って、必要な単位にインターフェースを分割
type Printer interface {
	Print(d Document)
}

// LSPに従って、必要な単位にインターフェースを分割
type Scanner interface {
	Scan(d Document)
}

// printer only
type MyPrinter struct {
	// ...
}

func (m MyPrinter) Print(d Document) {
	// ...
}

// インターフェースを組み合わせる
type Photocopier struct{}

func (p Photocopier) Scan(d Document) {
	//
}

func (p Photocopier) Print(d Document) {
	//
}

type MultiFunctionDevice interface {
	Printer
	Scanner
}

// interface combination + decorator
// 既存のインターフェースを組み合わせ、新しいインターフェースを作成することで
// 既存コードに変更を加えずに新しい機能を追加する
type MultiFunctionMachine struct {
	printer Printer
	scanner Scanner
}

func (m MultiFunctionMachine) Print(d Document) {
	m.printer.Print(d) // プリンター機能を移譲
}

func (m MultiFunctionMachine) Scan(d Document) {
	m.scanner.Scan(d) // スキャナー機能を移譲
}

func main() {

}

File: design_pattern/solid/isp/solid_isp.dart
class Document {
  final String name;
  final DateTime createdAt;

  Document({
    required this.name,
    DateTime? createdAt,
  }) : createdAt = createdAt ?? DateTime.now();
}

/// プリンター機能のための基本インターフェース
abstract interface class Printer {
  Future<void> print(Document document);
}

/// スキャナー機能のための基本インターフェース
abstract interface class Scanner {
  Future<Document> scan();
}

/// 複合機能のための基本インターフェース
abstract interface class MultiFunctionDevice implements Printer, Scanner {}

/// フォトコピー機の実装
class Photocopier implements MultiFunctionDevice {
  @override
  Future<void> print(Document document) async {
    try {
      // 印刷処理
      await Future.delayed(const Duration(seconds: 1));
    } catch (e) {
      throw PrinterException('印刷中にエラーが発生しました: $e');
    }
  }

  @override
  Future<Document> scan() async {
    try {
      // スキャン処理
      await Future.delayed(const Duration(seconds: 1));
      return Document(name: 'スキャン文書');
    } catch (e) {
      throw ScannerException('スキャン中にエラーが発生しました: $e');
    }
  }
}

/// デコレーターパターンを使用した実装
class MultiFunctionMachine implements MultiFunctionDevice {
  final Printer _printer;
  final Scanner _scanner;

  const MultiFunctionMachine({
    required Printer printer,
    required Scanner scanner,
  })  : _printer = printer,
        _scanner = scanner;

  @override
  Future<void> print(Document document) => _printer.print(document);

  @override
  Future<Document> scan() => _scanner.scan();
}

/// プリンター関連の例外
class PrinterException implements Exception {
  final String message;
  const PrinterException(this.message);
}

/// スキャナー関連の例外
class ScannerException implements Exception {
  final String message;
  const ScannerException(this.message);
}

Future<void> main() async {
  try {
    // 機器のインスタンス化
    final photocopier = Photocopier();
    final multiFunctionMachine = MultiFunctionMachine(
      printer: photocopier,
      scanner: photocopier,
    );

    // ドキュメントの作成
    final document = Document(name: 'テスト文書');

    print('=== フォトコピー機でのテスト ===');
    await photocopier.print(document);
    print('印刷完了');

    final scannedDoc = await photocopier.scan();
    print('スキャン完了: ${scannedDoc.name}');

    print('\n=== 複合機でのテスト ===');
    await multiFunctionMachine.print(document);
    print('印刷完了');

    final scannedByMFP = await multiFunctionMachine.scan();
    print('スキャン完了: ${scannedByMFP.name}');
  } on PrinterException catch (e) {
    print('プリンターエラー: ${e.message}');
  } on ScannerException catch (e) {
    print('スキャナーエラー: ${e.message}');
  } catch (e) {
    print('予期せぬエラー: $e');
  }
}
